Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -225,7 +302,29 @@ public struct HTTPHeaders: CustomStringConvertible {
     ///
     /// - Parameter name: The name of the header field to remove from the block.
     public mutating func remove(name: String) {
-        self.storage[name.lowercased()] = nil
+        guard !self.headers.isEmpty else {
+            return
+        }
+
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        var array: [Int] = []
+        // We scan from the back to the front so we can remove the subranges with as less overhead as possible.
+        for idx in stride(from: self.headers.count - 1, to: -1, by: -1) {
+            let header = self.headers[idx]
+            if stringLength == header.name.length && self.buffer.equalCaseInsensitiveASCII(seq: utf8, at: header.name) {
```

Comment:
If we add a proper length check into `equalCaseInsensitiveASCII` we can elide this length check here.